New: Add rule no-arrow-condition (fixes #3280)

lukekarrys opened the pull request:
This PR implements the `no-arrow-condition` rule as discussed in #3280.

I took the majority of the code from [this plugin](https://github.com/Jxck/eslint-plugin-arrow-function) which I had already modified slightly for my personal use case to add an option I'll describe below. I thought it would be easier to discuss in this PR rather than #3280 since the tests/docs do the best job of explaining what the option does.

By default this rule will warn about any case where the author uses `=>` but the intent might not be clear.

```js
/*eslint no-arrow-condition: 2*/
/*eslint-env es6*/

// All lines will produce warnings
if (a => 1) {}
while (a => 1) {}
for (var a = 1; a => 10; a++) {}
a => 1 ? 2 : 3
(a => 1) ? 2 : 3
var x = a => 1 ? 2 : 3
var x = (a) => 1 ? 2 : 3
```

I added an option `allowFunctionExpression` which when turned on, will not warn for implicit return arrow function expressions:

```js
/*eslint no-arrow-condition: [2, {"allowFunctionExpression": true}]*/
/*eslint-env es6*/

// Will not produce warnings
var x = a => 1 ? 2 : 3
var x = (a) => 1 ? 2 : 3
```

I wanted to open discussion for the option name, it's default value, and if other people agree it should be added. I added it because it matched my personal preference for writing implicit return arrow functions, but still wanting warnings for when an arrow function was used accidentally.

nzakas commented on docs/rules/no-arrow-condition.md:
```diff
@@ -0,0 +1,45 @@
+# Disallow arrow functions where a condition is expected (no-arrow-condition)
+
+Arrow functions (`=>`) are similar in syntax to some comparison operators (`>`, `<`, `<=`, and `>=`). This rule warns against using the arrow function syntax in places where a condition is expected. Even if the arguments of the arrow function are wrapped with parens, this rule still warns about it.
```
Can you add an example here of what a confusing usage might look like?

lukekarrys replied:
Yes I will.

nzakas commented:
Thanks, this looks like a good start. Since this code is taken from existing code, we need to include the complete license in the header comments of the files in order to be in adherence to the license. Can you add that?

Also, can you explain the `allowFunctionExpression` option? It doesn't seem to have anything to do with function expressions, and the two cases you highlight seem like they should definitely trigger a warning.

lukekarrys commented:
> Also, can you explain the allowFunctionExpression option?

It's an attempt to only warn on the code originally proposed in #3280 when an arrow function is mistakenly used inside an if/when/for statement. I called it that because the code will not warn when the arrow function is a function expression that is part of a variable assignment. I wasn't 100% happy with the name so I'm open to any alternatives.

The option makes it possible to only trigger warnings like the following:
```js
// Trigger a warning
for (var a = 1; a => 10; a++) {}

// Don't trigger a warning
var fn = arg => arg ? true : false
```

If the consensus is that the second piece of code above should always trigger a warning, then I'm fine for removing the option. I added it because it matched my personal preference of wanting warnings for the most obviously incorrect places where an arrow function was used accidentally, but not wanting to give up the ability to write implicit return arrow functions that contain conditionals.

lukekarrys commented:
> Since this code is taken from existing code, we need to include the complete license in the header comments of the files in order to be in adherence to the license.

Is just the text `MIT` enough, or would it need the full text of the MIT License?

nzakas commented:
Yeah, I'm not comfortable with that second option - I'm really not seeing the distinction you're trying to draw. Can you remove it? We can always see what feedback we get when it goes live. It's a lot easier to add a new option after it's released than it is to remove one.

You'll need the full text from the LICENSE file.

lukekarrys commented:
> Can you remove it? We can always see what feedback we get when it goes live. It's a lot easier to add a new option after it's released than it is to remove one.

Yep, sounds good. I'll remove it.

>  I'm really not seeing the distinction you're trying to draw.

Just to try and show the distinction (since I'm definitely assuming something somewhere) I was trying to separate cases where an arrow function is definitely not intended vs just ambiguous. Am I incorrectly assuming that there are two different levels of errors here? I'll definitely remove the option, I'm just curious where I was making an incorrect assumption.

```js
// You would never want this
for (var a = 1; a => 10; a++) {}

// You might want this sometimes
var fn = arg => arg ? true : false
// even though the intent could be
var fn = function (arg) { return arg ? true : false }
// or the intent could be
var fn = arg >= arg ? true : false
```

nzakas commented:
Ah, I see what you mean now, thanks for explaining. I still think we're better off without that option to start. People can rewrite this:

```js
var fn = arg => arg ? true : false;
```

As

```js
var fn = arg => { return arg ? true : false };
```

Which I'd argue makes it a lot easier to understand what's going on.

lukekarrys commented:
- [x] Added the license to the rule and test file
- [x] Removed the code and documentation for the option
- [x] Added example to documentation highlighting confusing usage

lukekarrys commented:
Repushed to fix issue with fenced code blocks missing surrounding blank lines.

Jxck commented:
FYI https://github.com/Jxck/eslint-plugin-arrow-function/pull/5

nzakas commented:
LGTM. Thanks!
